tests/unit/vector/shufflevector1.opt

```diff
@@ -48,10 +48,28 @@ Name: constant-folding-8
 Name: constant-folding-9
 %r = shufflevector <2 x i4> <i4 3, i4 5>, <2 x i4> <i4 4, i4 6>, <3 x i32> <i32 poison, i32 1, i32 4>
   =>
-%r = <3 x i4> <i4 poison, i4 5, i4 poison>
+%r = <3 x i4> <i4 poison, i4 5, i4 undef>
 
 
 Name: constant-folding-10
 %r = shufflevector <2 x float> <float 3.0, float 5.0>, <2 x float> <float 4.0, float 6.0>, <3 x i32> <i32 poison, i32 1, i32 4>
   =>
-%r = <3 x float> <float poison, float 5.0, float poison>
+%r = <3 x float> <float poison, float 5.0, float undef>
+
+
+Name: constant-folding-11
+%r = shufflevector <2 x float> <float 3.0, float 5.0>, <2 x float> <float 4.0, float 6.0>, <3 x i32> <i32 poison, i32 undef, i32 undef>
+  =>
+%r = <3 x float> <float poison, float undef, float undef>
+
+
+Name: constant-folding-12
+%r = shufflevector <2 x float> <float poison, float 5.0>, <2 x float> <float 4.0, float 6.0>, <4 x i32> <i32 5, i32 undef, i32 0, i32 1>
+  =>
+%r = <4 x float> <float undef, float undef, float poison, float 5.0>
+
+
+Name: constant-folding-13
```

Comment:
This test case is failed because undef mask shufflevector still results in poison.

Reply:
Ah damn.
When the mask is undef, it can take any value. In particular, if it's 0, it picks the poison argument. So we conclude the result can be either poison, undef, or any of the other arguments.
That's the problem with these semantics that "detect" undef and give some value for that case. That usually doesn't work, and here's another example.
Do you want to try to produce an end-to-end miscompilation using this information? (you should be able to use the same gadgets I've used for the other transformation).
You can also run this patch through the LLVM test suite and see what comes out. You may just find the miscompilation there :)

Reply:
I need some more time to figure out a miscompilation :)

Reply:
@nunoplopes I have tried to figure out a miscompilation. But for now I think it is impossible after Sanjay's patch.

In order to generate a case containing (undef => poison), I tried to trick the optimizer to synthesize a shufflevector instruction from undef, and then expect some  (undef => normal value) optimization to rewrite the undef in the mask to point to an poison value. See https://godbolt.org/z/E8qFwk for an non-working example.

The reason is that for shufflevector, except for the identity simplification optimization discussed in the bugzilla, there is no other transformations happens on the mask of shufflevector. In the other words, after the (undef => shufflevector %v, undef, undef) transformation is done, there is no other writes on the mask, the poison inside %v will never get exposed.

Reply:
I've read a few tests in the InstCombine test suite to find out if any other rewrite that turns undef into normal values so that I can use it to composite a miscompile, but I cannot find one. I've also runned the alive-tv on test suite with this patch again, the results are the same.

Reply:
Fixing that identity simplification optimization may make the compilation sound for a moment. But in the longer term this undef mask semantic may not be the best choice. As to make the translation sound, most rewrites on undef masks may not fit.